pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,549 @@ package service
 
 import (
 	"context"
+	"math"
 	"reflect"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
 
+	"github.com/vdaas/vald/internal/errgroup"
 	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/k8s/configmap"
+	"github.com/vdaas/vald/internal/k8s/decoder"
+	"github.com/vdaas/vald/internal/k8s/job"
+	mpod "github.com/vdaas/vald/internal/k8s/metrics/pod"
+	"github.com/vdaas/vald/internal/k8s/pod"
+	"github.com/vdaas/vald/internal/k8s/statefulset"
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/safety"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/config"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/model"
 )
 
+const (
+	jobType = "rebalancer"
+)
+
+// Rebalancer represents the rebalancer interface.
 type Rebalancer interface {
-	Start(context.Context) (<-chan error, error)
+	Start(ctx context.Context) (<-chan error, error)
 }
 
-type rebalancer struct{}
+type rebalancer struct {
+	podName      string
+	podNamespace string
+
+	jobs           atomic.Value
+	jobName        string
+	jobNamespace   string
+	jobTemplate    atomic.Value
+	jobTemplateKey string // config map key of job template
+
+	agentName         string
+	agentNamespace    string
+	agentResourceType config.AgentResourceType
+	pods              atomic.Value
+	podMetrics        atomic.Value
+
+	jobConfigmapName      string
+	jobConfigmapNamespace string
+
+	statefulSets atomic.Value
+
+	leaderElectionID string
 
-func New(opts ...Option) (dsc Rebalancer, err error) {
+	rcd           time.Duration // reconcile check duration
+	eg            errgroup.Group
+	ctrl          k8s.Controller
+	tolerance     float64
+	rateThreshold float64
+	decoder       *decoder.Decoder
+}
+
+// NewRebalancer initialize job, configmap, pod, podMetrics, statefulset, replicaset, daemonset reconciler.
+// And it returns the rebalancer implemenation or any error occurred.
+func NewRebalancer(opts ...RebalancerOption) (Rebalancer, error) {
 	r := new(rebalancer)
-	for _, opt := range append(defaultOpts, opts...) {
+
+	for _, opt := range append(defaultRebalancerOpts, opts...) {
 		if err := opt(r); err != nil {
 			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
 		}
 	}
+
+	job, err := job.New(
+		job.WithControllerName("job rebalancer"),
+		job.WithNamespaces(r.jobNamespace),
+		job.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		job.WithOnReconcileFunc(func(jobList map[string][]job.Job) {
+			log.Debugf("[reconcile Job] length Joblist: %d", len(jobList))
+			jobs, ok := jobList[r.jobName]
+			if ok {
+				r.jobs.Store(jobs)
+			} else {
+				r.jobs.Store(make([]job.Job, 0))
```

Comment:
Should we prevent re-allocate? 🤔 
Should we load data from cache and padding slice and store it?🤔 
if I made a mistake, I am sorry.

Reply:
When job has two or more jobs, how does your plant do?